# Задание: 
Для каждого кадра с камеры 1 нужно найти соответствующий (ближайший по времени) кадр с камеры 2. 
* Камеры записывают в формате (номер кадра, время его захвата) ~ (frame, timestamp).
* Камеры могут делать запись с разной частотой кадров в секунду и наличием шума (несоответствии времени кадра ожидаемому времени).

## Настройки/Гиперпараметры/Импорты

In [1]:
import time # для подсчёта времени работы
import numpy as np # для быстрой работы с массивами
from multiprocessing import Pool # для параллельного запуска (нужны процессы, а не потоки из multithreading/concurrent.futures)
from numba import njit, prange # импорт компилятора для ускорения питона (prange для параллелизации)
from numba import float64, int32 # для создания сигнатур (почему-то крашится, поэтому в коде использование закоменчено)

In [2]:
cores_number = 2 # число ядер, на которые будет распараллеливаться выполнение кода
runs = 5 # число запусков для усреднения времени

## Вспомогательные функции

In [3]:
def make_timestamps(fps: int, st_ts: float, fn_ts: float) -> np.ndarray: # функция генерации записей с камеры
    """
    Создаёт np.array timestamps (массив времени записи кадра). 
    Этот массив дискретизирован по fps, но не равномерно.
    Возвращаемые timestamps отсортированы и уникальны.\n
    Parameters:
        * fps: среднее число кадров в секунду
        * st_ts: первый timestamp в последовательности (время первого кадра)
        * fn_ts: первый timestamp в последовательности (время последнего кадра) \n
    Returns:
        * np.ndarray: сгенерированные timestamps (массивы времени записи кадра)
    """
    timestamps = np.linspace(st_ts, fn_ts, int((fn_ts - st_ts) * fps)) # создаём временные записи из равномерного распределения
    timestamps += np.random.randn(len(timestamps)) # зашумляем кадры
    timestamps = np.unique(np.sort(timestamps)) # сортируем кадры по времени и оставляем только уникальные
    return timestamps

In [4]:
def calc_average_time(alg, data, runs: int=100) -> (float, np.ndarray): # функция для подсчёта среднего времени работы алгоритма
    """
    Функция для подсчёта среднего времени работы алгоритма.\n
    Parameters:
        * alg: алгоритм для тестирования
        * data: данные на вход алгоритма
        * runs: число запусков для усреднения времени
    Returns:
        * float: среднее время работы
        * np.ndarray: получившийся ответ
    """
    time_start = time.perf_counter() # замеряем время начала
    for i in range(runs): # делаем runs запусков
        answer = alg(*data) # вызываем алгоритм (*data для развёртывания переданных данных для алгоритма, если они в list-like формате)
        print(i, (time.perf_counter() - time_start)/(i+1)) # вывод среднего времени i-ой итерации (i+1 из-за нумерации с нуля)
    time_average = (time.perf_counter() - time_start) / runs # считаем среднее время
    return time_average, answer # возвращаем среднее время и получившийся ответ

In [5]:
# generate timestamps for the first camera
# timestamps1 = make_timestamps(30, time.time() - 100, time.time() + 3600 * 2)
# generate timestamps for the second camera
# timestamps2 = make_timestamps(60, time.time() + 200, time.time() + 3600 * 2.5)

timestamps1 = make_timestamps(30, time.time() - 100, time.time() + 900 * 2) # не 3600 (60 секунд * 60 минут), а лишь (15 * 60) = 15 минут
timestamps2 = make_timestamps(60, time.time() + 200, time.time() + 900 * 2.5) # не 3600 (60 секунд * 60 минут), а лишь (15 * 60) = 15 минут

In [6]:
timestamps1.shape

(56998,)

## Варианты решения

### 1) Глупо считаем разницу во времени между i-м кадром с первой камеры и всеми кадрами со второй камеры. В ответ записываем номер кадра со второй камеры с наименьшим модулем разности времени.

In [7]:
@njit #====================== использовать ли numba компилятор для функции ============================
def match_timestamps_naive(timestamps1: np.ndarray, timestamps2: np.ndarray) -> np.ndarray: 
    """
    * Делаем для каждого элемента первого массива: - O(n)
            * из всех элементов второго массива вычитается элемент первого массива (считаем delta) за O(2n) ~ O(n)
            * в получившихся delta ищем наименьший элемент за O(2n) ~ O(n)\n
    Итоговая сложность: O(размер_первого_массива * (размер_второго_массива + размер_второго_массива)) = O(n*(2n+2n)) = O(4n^2) ~ O(n^2)
    Parameters:
        * timestamps1: массив с временами с камеры 1
        * timestamps2: массив с временами с камеры 2\n
    Returns:
        * np.ndarray: массив соответствий кадров с камеры 1 к кадрам камеры 2
    """
    
    # так как в задании первая камера пишет в ~30 fps -> размер первого массива n; вторая камера пишет в ~60 fps -> размер второго массива не сильно больше 2n; сокращаем на константы и получаем n
    frames_count = timestamps1.shape[0] # число кадров на камере 1
    correspondence = np.zeros(frames_count, dtype=np.int32) # создаём массив под номера кадров с типом int32 для уменьшения потребляемой памяти

    #===================== вариант с циклом в np.arange ================================================
    # for frame in np.arange(frames_count): # идём по номерам кадров с первой камеры
    #--------------------- вариант с циклом в enumerate ------------------------------------------------
    for frame, frame_time in enumerate(timestamps1): # идём по кадрам
    #===================================================================================================
        #===================== вариант с сохранением delta =============================================
        # deltas = np.absolute(timestamps2 - timestamps1[frame]) # считаем разницу между временами
        # correspondence[frame] = deltas.argmin() # берём номер кадра с наименьшей разницей
        #--------------------- вариант без сохранения delta --------------------------------------------
        correspondence[frame] = np.absolute(timestamps2 - frame_time).argmin() # цикл с enumerate
        # correspondence[frame] = np.absolute(timestamps2 - timestamps1[frame]).argmin() # цикл с np.arange
        #===============================================================================================

    return correspondence

In [8]:
true_time, true_answer = calc_average_time(match_timestamps_naive, (timestamps1, timestamps2), runs=runs) # njit, цикл в enumerate

0 24.950418900000386
1 20.281998550000026
2 19.04419333333347
3 18.18515685
4 17.745648980000077


<pre>
- naive no njit + np.arange ~ 20.0 - 70.0 секунд
- naive no njit + enumerate ~ 18.0 - 19.0 секунд
- naive njit    + np.arange ~ 15.5 - 15.8 секунд
- naive njit    + enumerate ~ 15.0 - 15.7 секунд
</pre>

### 2) Параллельная версия первого алгоритма

In [9]:
from funcs import match_timestamps_naive # импорт функции, так как в notebook multiprocessing навсегда зависает при вызове функции из этого же notebook (???)

In [10]:
def match_timestamps_naive_parallel(timestamps1: np.ndarray, timestamps2: np.ndarray, func, cores_number: np.int8) -> np.ndarray: 
    """
    * Делаем для каждого элемента первого массива с распараллеливанием: - O(n/cores_number)
            * из всех элементов второго массива вычитается элемент первого массива (считаем delta) за O(2n) ~ O(n)
            * в получившихся delta ищем наименьший элемент за O(2n) ~ O(n)\n
    Итоговая сложность: O(размер_первого_массива / число_ядер * (размер_второго_массива + размер_второго_массива)) = O(n/cores_number*(2n+2n)) = O(4/cores_number*n^2) ~ O(n^2/cores_number)\n
    Parameters:
        * timestamps1: массив с временами с камеры 1
        * timestamps2: массив с временами с камеры 2
        * func: запускаемая в параллель функция
        * cores_number: число ядер, на которое будет распараллеливание\n
    Returns:
        * np.ndarray: массив соответствий кадров с камеры 1 к кадрам камеры 2
    """

    # так как в задании первая камера пишет в ~30 fps -> размер первого массива n; вторая камера пишет в ~60 fps -> размер второго массива 2n; сокращаем на константы и получаем n
    frames_count = timestamps1.shape[0] # число кадров с первой камеры
    with Pool(cores_number) as p: # запускаем cores_number процессов
        correspondence = np.concatenate(p.starmap(func, [(timestamps1[int(i*frames_count/cores_number):int((i+1)*frames_count/cores_number)], timestamps2) for i in np.arange(cores_number)]), dtype=np.int32)
        #  - каждому процессу отправляем равную часть кадров с первой камеры и все кадры со второй
        #  - конкатенируем вернувшиеся из процессов массивы с индексами
        # процесс вызывает переданную функцию func с параметрами: срез кадров с первой камеры, все кадры второй камеры
        # не обязательно передавать chunksize, так как распараллеливание уже идёт по кейсам/инпутам для функций
        # dtype=np.int32 для уменьшения потребляемой памяти и ускорения работы

    return correspondence

In [11]:
time_naive_parallel, answer_naive_parallel = calc_average_time(match_timestamps_naive_parallel, (timestamps1, timestamps2, match_timestamps_naive, cores_number), runs=runs) # подсчёт среднего времени (с njit и enumerate циклом)

0 10.85389529999975
1 11.15200354999979
2 11.337617466666567
3 11.252855424999893
4 11.30571489999993


In [12]:
np.all(answer_naive_parallel==true_answer) # проверка, что параллельная функция отработала корректно

True

<pre>
- parallel naive no njit + np.arange ~ 40.0 - 41.0 секунд
- parallel naive no njit + enumerate ~ 41.0 - 43.0 секунд
- parallel naive njit    + np.arange ~ 11.2 - 12.0 секунд
- parallel naive njit    + enumerate ~ 10.0 - 11.3 секунд
</pre>

### 3) Идём по последовательности кадров с первой камеры, запоминанием найденное соответствие со второй (номер delta, после которого разница во времени начинают увеличиваться). На следующем шаге по кадрам на второй камере начинаем идти с найденного соответствия на предыдущем шаге (так как кадры отсортированы в порядке неубывания - нет смысла рассматривать кадры, что заведомо имеют большую delta из-за увеличения времени на первой камере).

In [13]:
# @njit(float64[:](float64[:], float64[:], int32)) # использовать ли numba компилятор для функции с заданием сигнатуры для изначальной компиляции
@njit #====================== использовать ли numba компилятор для функции ============================
def match_timestamps_iterated(timestamps1: np.ndarray, timestamps2: np.ndarray, cam2_frame: np.int32=0) -> np.ndarray: 
    """
    * Ищем начальный кадр для второй камеры - O(2n), однако реальная сложность меньше, так как cam2_frame уже найден эвристически с учётом числа кадров на первой и второй камере
    * Делаем для каждого элемента (времени) с первой камеры: - O(n)
            * Вычитаем из соответствующего времени кадра второй камеры значение времени кадра с первой камеры - O(2n) в худшем случае, O(2) ~ O(1) в среднем, так как кадры распределены равномерно и во втором массиве не на порядок больше элементов
                    * если дельта уменьшилась - переходим к следующему кадру на второй камере и так далее, пока дельта уменьшается\n
                    * если дельта не уменьшилась - возвращаем предыдущий кадр, как самый близкий по времени\n
    Итоговая сложность: O(размер_второго_массива + размер_первого_массива * размер_второго_массива) ~ [в среднем] ~ O(размер_второго_массива + размер_первого_массива * 1) ~ O(n)\n
    Parameters:
        * timestamps1: массив с временами с камеры 1
        * timestamps2: массив с временами с камеры 2
        * cam2_frame: с какого элемента рассматривать кадры с камеры 2 (важен при параллельном запуске)\n
    Returns:
        * np.ndarray: массив соответствий кадров с камеры 1 к кадрам камеры 2
    """
    
    # так как в задании первая камера пишет в ~30 fps -> размер первого массива n; вторая камера пишет в ~60 fps -> размер второго массива не сильно больше 2n; сокращаем на константы и получаем n
    frames_count = timestamps1.shape[0] # число кадров на камере 1
    max_frame = timestamps2.shape[0] - 1 # максимальный номер кадра, что может соответствовать кадру с первой камеры (-1 из-за индексации с нуля)
    correspondence = np.zeros(frames_count, dtype=np.int32) # создаём массив под номера кадров с типом int32 для уменьшения потребляемой памяти

    if 0 > cam2_frame or cam2_frame > max_frame: # защита от дурака с cam2_frame
        cam2_frame = 0 # зануление начального кадра второй камеры

    # подбираем кадр на второй камере, с которого будем начинать (эвристически уже передали ожидаемый cam2_frame, нужно его лишь правильно сместить, если он "обогнал" кадры с камеры 1)
    delta = np.abs(timestamps2[cam2_frame] - timestamps1[0]) # текущая разность времени
    while np.abs(timestamps2[max(0, cam2_frame-1)] - timestamps1[0]) < delta: # если дельта уменьшается при предыдущем кадре, то переходим на него (условие — строго меньше, так как времена всех кадров уникальны)
        cam2_frame -= 1 # переходим на предыдущий кадр (без max, так как при нуле мы бы сюда не зашли в цикл)
        delta = np.abs(timestamps2[cam2_frame] - timestamps1[0]) # обновляем текущее значение delta (без max, так как при нуле мы бы не зашли в цикл)
    # после этого цикла мы либо находимся на оптимальном соответствии первого кадра первой камеры с кадром со второй камеры, либо оптимум — далее по времени (но не раньше!)

    #===================== вариант с циклом в np.arange ================================================
    # for frame in np.arange(frames_count): # идём по номерам кадров с первой камеры
    #     delta = np.abs(timestamps2[cam2_frame] - timestamps1[frame]) # текущая разность времени
    #     while np.abs(timestamps2[min(cam2_frame+1, max_frame)] - timestamps1[frame]) < delta: # если при переходе на следующий кадр дельта (разница времени) уменьшилась, то обновляем delta и соответствующий кадр на второй камере 
    #         cam2_frame += 1 # переходим на следующий кадр
    #         delta = np.abs(timestamps2[cam2_frame] - timestamps1[frame])  # обновляем delta
    #     correspondence[frame] = cam2_frame # если дельта перестала уменьшаеться — записываем найденный кадр
    #--------------------- вариант с циклом в enumerate ------------------------------------------------
    for frame, frame_time in enumerate(timestamps1): # идём по кадрам (frame - номер кадра на первой камере, frame_time - время его запечатления)
        delta = np.abs(timestamps2[cam2_frame] - frame_time) # текущая разность времени
        while np.abs(timestamps2[min(cam2_frame+1, max_frame)] - frame_time) < delta: # если при переходе на следующий кадр дельта (разница времени) уменьшилась, то обновляем delta и соответствующий кадр на второй камере 
            cam2_frame += 1 # переходим на следующий кадр
            delta = np.abs(timestamps2[cam2_frame] - frame_time)  # обновляем delta
        correspondence[frame] = cam2_frame # если дельта перестала уменьшаеться — записываем найденный кадр
    #===============================================================================================

    return correspondence

In [14]:
time_iterated, answer_iterated = calc_average_time(match_timestamps_iterated, (timestamps1, timestamps2), runs=runs) # подсчёт среднего времени (с njit и enumerate циклом)
# время сильно разнится из-за создания бинаря для функции с njit при первом запуске (последующие запуски идут всегда быстрее)

0 0.5652193999999326
1 0.2838906499998757
2 0.18977626666658884
3 0.14298617500003274
4 0.1151557999999568


In [15]:
np.all(answer_iterated==true_answer) # проверка, что функция отработала корректно

True

<pre>
- iterated no njit + np.arange ~ 0.7451 секунд
- iterated no njit + enumerate ~ 0.6525 секунд
- iterated njit    + np.arange ~ 0.0882 секунд
- iterated njit    + enumerate ~ 0.0795 секунд
</pre>

### 4) Параллельная реализация третьего алгоритма

In [16]:
from funcs import match_timestamps_iterated # импорт функции, так как в notebook multiprocessing навсегда зависает при вызове функции из этого же notebook (???)

In [17]:
def match_timestamps_iterated_parallel(timestamps1: np.ndarray, timestamps2: np.ndarray, func, cores_number: np.int8) -> np.ndarray: 
    """
    * Ищем начальный кадр для второй камеры - O(2n), однако реальная сложность меньше, так как cam2_frame уже найден эвристически с учётом числа кадров на первой и второй камере
    * Делаем для каждого элемента первого массива (времени) с распараллеливанием: - O(n/cores_number)
            * Вычитаем из соответствующего времени кадра второй камеры значение времени кадра с первой камеры - O(2n) в худшем случае, O(2) ~ O(1) в среднем, так как кадры распределены равномерно и во втором массиве не на порядок больше элементов
                    * если дельта уменьшилась - переходим к следующему кадру на второй камере и так далее, пока дельта уменьшается
                    * если дельта не уменьшилась - возвращаем предыдущий кадр, как самый близкий по времени\n
    Итоговая сложность: O(размер_второго_массива + размер_первого_массива / число_ядер * размер_второго_массива) ~ [в среднем] ~ O(размер_второго_массива + размер_первого_массива / число_ядер * 1) ~ O(n/cores_number)\n
    Parameters:
        * timestamps1: массив с временами с камеры 1
        * timestamps2: массив с временами с камеры 2
        * func: запускаемая в параллель функция
        * cores_number: число ядер, на которое будет распараллеливание\n
    Returns:
        * np.ndarray: массив соответствий кадров с камеры 1 к кадрам камеры 2
    """

    # так как в задании первая камера пишет в ~30 fps -> размер первого массива n; вторая камера пишет в ~60 fps -> размер второго массива 2n; сокращаем на константы и получаем n
    frames_count = timestamps1.shape[0] # число кадров с первой камеры
    with Pool(cores_number) as p: # запускаем cores_number процессов
        correspondence = np.concatenate(p.starmap(func, [(timestamps1[int(i*frames_count/cores_number):int((i+1)*frames_count/cores_number)], timestamps2, int(2*i*frames_count/cores_number)) for i in np.arange(cores_number)]), dtype=np.int32)
        #  - каждому процессу отправляем равную часть кадров с первой камеры и все кадры со второй, а также ожидаемый номер соответствия первого кадра на камере 1 с камерой 2
        #  - конкатенируем вернувшиеся из процессов массивы с индексами
        # процесс вызывает переданную функцию func с параметрами: срез кадров с первой камеры, все кадры второй камеры, а также ожидаемый номер соответствия первого кадра на камере 1 с камерой 2
        # не обязательно передавать chunksize, так как распараллеливание уже идёт по кейсам/инпутам для функций
        # dtype=np.int32 для уменьшения потребляемой памяти и ускорения работы

    return correspondence

In [18]:
time_iterated_parallel, answer_iterated_parallel = calc_average_time(match_timestamps_iterated_parallel, (timestamps1, timestamps2, match_timestamps_iterated, cores_number), runs=runs) # подсчёт среднего времени (с njit и enumerate циклом)

0 2.9934794000000693
1 3.0763882999999623
2 2.9628640333333656
3 2.8483673499999895
4 2.729524000000038


In [19]:
np.all(answer_iterated_parallel==true_answer) # проверка, что функция отработала корректно

True

<pre>
- parallel iterated no njit + np.arange ~ 1.3736 секунд
- parallel iterated no njit + enumerate ~ 1.3057 секунд
- parallel iterated njit    + np.arange ~ 2.4365 секунд
- parallel iterated njit    + enumerate ~ 2.1035 секунд
</pre>

### 5*) Magnum Opus — немного переосмысленный третий алгоритм

In [23]:
@njit # использование компилятора
def match_timestamps_iterated_v2(timestamps1: np.array, timestamps2: np.array, cores_number: np.int8=1) -> np.array:
    """
    * Ищем начальный кадр для второй камеры - O(2n), однако реальная сложность меньше, так как cam2_frame уже найден эвристически с учётом числа кадров на первой и второй камере
    * Делаем для каждого элемента (времени) с первой камеры: - O(n)
            * Вычитаем из соответствующего времени кадра второй камеры значение времени кадра с первой камеры - O(2n) в худшем случае, O(2) ~ O(1) в среднем, так как кадры распределены равномерно и во втором массиве не на порядок больше элементов
                    * если дельта уменьшилась - переходим к следующему кадру на второй камере и так далее, пока дельта уменьшается\n
                    * если дельта не уменьшилась - возвращаем предыдущий кадр, как самый близкий по времени\n
    Итоговая сложность: O(размер_второго_массива + размер_первого_массива * размер_второго_массива) ~ [в среднем] ~ O(размер_второго_массива + размер_первого_массива * 1) ~ O(n)\n
    Parameters:
        * timestamps1: массив с временами с камеры 1
        * timestamps2: массив с временами с камеры 2
        * cores_number: число ядер для распараллеливания\n
    Returns:
        * np.ndarray: массив соответствий кадров с камеры 1 к кадрам камеры 2
    """
    frames_count = timestamps1.shape[0] # число кадров на камере 1
    correspondence = np.full(shape=frames_count, fill_value=-1, dtype=np.int32) # создаём массив под номера кадров с типом int32 для уменьшения потребляемой памяти
    max_frame = timestamps2.shape[0] - 1  # максимальный номер кадра, что может соответствовать кадру с первой камеры (-1 из-за индексации с нуля)
    for proc_id in prange(cores_number): # разделяем задачу на проессы
        t1_start = int(proc_id*frames_count/cores_number) # # с какого элемента рассматриваем кадры с первой камеры 
        t1_end = int((proc_id+1)*frames_count/cores_number) # по какой элемент рассматриваем кадры с первой камеры (без -1, так как slice не включает последний элемент)
        cam2_frame = int(2*proc_id*frames_count/cores_number) # с какого элемента рассматривать кадры с камеры 2 (важен при параллельном запуске)
        
        if 0 > cam2_frame or cam2_frame > max_frame: # защита от дурака с cam2_frame
            cam2_frame = 0 # зануление начального кадра второй камеры

        # подбираем кадр на второй камере, с которого будем начинать (эвристически уже передали ожидаемый cam2_frame, нужно его лишь правильно сместить, если он "обогнал" кадры с камеры 1)
        delta = np.abs(timestamps2[cam2_frame] - timestamps1[t1_start]) # текущая разность времени
        while np.abs(timestamps2[max(0, cam2_frame-1)] - timestamps1[t1_start]) < delta: # если дельта уменьшается при предыдущем кадре, то переходим на него (условие — строго меньше, так как времена всех кадров уникальны)
            cam2_frame -= 1 # переходим на предыдущий кадр (без max, так как при нуле мы бы сюда не зашли в цикл)
            delta = np.abs(timestamps2[cam2_frame] - timestamps1[t1_start]) # обновляем текущее значение delta
        # после этого цикла мы либо находимся на оптимальном соответствии первого кадра первой камеры с кадром со второй камеры, либо оптимум — далее по времени (но не раньше!)

        t1_current = t1_start # текущий кадр
        for _, frame_time in enumerate(timestamps1[t1_start:t1_end]): # идём по кадрам (_ - номер кадра на первой камере, frame_time - время его запечатления)
            delta = np.abs(timestamps2[cam2_frame] - frame_time) # текущая разность времени
            while np.abs(timestamps2[min(cam2_frame+1, max_frame)] - frame_time) < delta: # если при переходе на следующий кадр дельта (разница времени) уменьшилась, то обновляем delta и соответствующий кадр на второй камере 
                cam2_frame += 1 # переходим на следующий кадр
                delta = np.abs(timestamps2[cam2_frame] - frame_time)  # обновляем delta
            correspondence[t1_current] = cam2_frame # если дельта перестала уменьшаеться — записываем найденный кадр с учётом сдвига от параллельного выполнения
            t1_current += 1 # изменяем текущий кадр
    return correspondence

In [24]:
time_iterated_v2, answer_iterated_v2 = calc_average_time(match_timestamps_iterated_v2, (timestamps1, timestamps2, 1), runs=runs) # подсчёт среднего времени (с njit и enumerate циклом)

0 0.4678132999997615
1 0.23447880000003352
2 0.156625333333371
3 0.11768195000001924
4 0.09431329999997615


In [26]:
np.all(answer_iterated_v2==true_answer) # проверка, что функция отработала корректно

True

<pre>
- iterated v2 ~ 0.0943 - 0.4678 секунд
</pre>

### 6*) Параллельная реализация пятого алгоритма

In [27]:
@njit(parallel=True) # использование компилятора с параллельным запуском
def match_timestamps_iterated_v2_parallel(timestamps1: np.array, timestamps2: np.array, cores_number: np.int8) -> np.array:
    """
    * Ищем начальный кадр для второй камеры - O(2n), однако реальная сложность меньше, так как cam2_frame уже найден эвристически с учётом числа кадров на первой и второй камере
    * Делаем для каждого элемента (времени) с первой камеры: - O(n/cores_number)
            * Вычитаем из соответствующего времени кадра второй камеры значение времени кадра с первой камеры - O(2n) в худшем случае, O(2) ~ O(1) в среднем, так как кадры распределены равномерно и во втором массиве не на порядок больше элементов
                    * если дельта уменьшилась - переходим к следующему кадру на второй камере и так далее, пока дельта уменьшается\n
                    * если дельта не уменьшилась - возвращаем предыдущий кадр, как самый близкий по времени\n
    Итоговая сложность: O(размер_второго_массива + размер_первого_массива / число ядер * размер_второго_массива) ~ [в среднем] ~ O(размер_второго_массива + размер_первого_массива / число ядер * 1) ~ O(n/cores_number)\n
    Parameters:
        * timestamps1: массив с временами с камеры 1
        * timestamps2: массив с временами с камеры 2
        * cores_number: число ядер для распараллеливания\n
    Returns:
        * np.ndarray: массив соответствий кадров с камеры 1 к кадрам камеры 2
    """
    frames_count = timestamps1.shape[0] # число кадров на камере 1
    correspondence = np.full(shape=frames_count, fill_value=-1, dtype=np.int32) # создаём массив под номера кадров с типом int32 для уменьшения потребляемой памяти
    max_frame = timestamps2.shape[0] - 1  # максимальный номер кадра, что может соответствовать кадру с первой камеры (-1 из-за индексации с нуля)
    for proc_id in prange(cores_number): # разделяем задачу на проессы
        t1_start = int(proc_id*frames_count/cores_number) # # с какого элемента рассматриваем кадры с первой камеры 
        t1_end = int((proc_id+1)*frames_count/cores_number) # по какой элемент рассматриваем кадры с первой камеры (без -1, так как slice не включает последний элемент)
        cam2_frame = int(2*proc_id*frames_count/cores_number) # с какого элемента рассматривать кадры с камеры 2 (важен при параллельном запуске)
        
        if 0 > cam2_frame or cam2_frame > max_frame: # защита от дурака с cam2_frame
            cam2_frame = 0 # зануление начального кадра второй камеры

        # подбираем кадр на второй камере, с которого будем начинать (эвристически уже передали ожидаемый cam2_frame, нужно его лишь правильно сместить, если он "обогнал" кадры с камеры 1)
        delta = np.abs(timestamps2[cam2_frame] - timestamps1[t1_start]) # текущая разность времени
        while np.abs(timestamps2[max(0, cam2_frame-1)] - timestamps1[t1_start]) < delta: # если дельта уменьшается при предыдущем кадре, то переходим на него (условие — строго меньше, так как времена всех кадров уникальны)
            cam2_frame -= 1 # переходим на предыдущий кадр (без max, так как при нуле мы бы сюда не зашли в цикл)
            delta = np.abs(timestamps2[cam2_frame] - timestamps1[t1_start]) # обновляем текущее значение delta
        # после этого цикла мы либо находимся на оптимальном соответствии первого кадра первой камеры с кадром со второй камеры, либо оптимум — далее по времени (но не раньше!)

        t1_current = t1_start # текущий кадр
        for _, frame_time in enumerate(timestamps1[t1_start:t1_end]): # идём по кадрам (_ - номер кадра на первой камере, frame_time - время его запечатления)
            delta = np.abs(timestamps2[cam2_frame] - frame_time) # текущая разность времени
            while np.abs(timestamps2[min(cam2_frame+1, max_frame)] - frame_time) < delta: # если при переходе на следующий кадр дельта (разница времени) уменьшилась, то обновляем delta и соответствующий кадр на второй камере 
                cam2_frame += 1 # переходим на следующий кадр
                delta = np.abs(timestamps2[cam2_frame] - frame_time)  # обновляем delta
            correspondence[t1_current] = cam2_frame # если дельта перестала уменьшаеться — записываем найденный кадр с учётом сдвига от параллельного выполнения
            t1_current += 1 # изменяем текущий кадр
    return correspondence

In [28]:
time_iterated_v2_parallel, answer_iterated_v2_parallel = calc_average_time(match_timestamps_iterated_v2_parallel, (timestamps1, timestamps2, cores_number), runs=runs) # подсчёт среднего времени (с njit и enumerate циклом)

0 1.3647298000000774
1 0.6832738499999778
2 0.45581383333334696
3 0.3420914499999981
4 0.2738818800000445


In [29]:
np.all(answer_iterated_v2_parallel==true_answer) # проверка, что функция отработала корректно

True

<pre>
- parallel iterated v2 ~ 0.2738 - 1.3647 секунд
</pre>

# Выводы

* Лучшим вариантом, как и ожидалось, оказался итерационный алгоритм (njit + enumerate).
* Циклы с enumerate показали себя лучше, чем циклы с np.arange.
* Параллелизация может ускорить выполнение кода, однако её эффективность зависит от изначальной времени работы (возможен случай, что на распараллеливание уходит больше времени, чем на сам подсчёт).
* njit ускоряет код вплоть x10, так как компилирует бинарные файлы для функций (без сигнатуры бинарный файл создастся при первом вызове, а не при определении функции), однако иногда он может наоборот — замедлить (см вариант 4 с параллельным запуском).